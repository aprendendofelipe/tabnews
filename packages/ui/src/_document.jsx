import NextDocument, { Head, Html, Main, NextScript } from 'next/document.js';
import Script from 'next/script.js';
import { ServerStyleSheet } from 'styled-components';

// Script related to `AutoThemeProvider`
export const noFlashScript = `if (['auto','night','dark','day','light'].includes(localStorage.getItem('colorMode')))
document.documentElement.setAttribute('data-no-flash', true)`;

let documentConfig = {
  htmlProps: {},
};

/**
 * Configure the document properties.
 * @param {Object} config - The configuration object.
 * @param {Object} [config.htmlProps] - The HTML properties to be set on the <html> tag.
 * @param {import("react").ReactNode} [config.headChildren] - The children to be added to the <head> tag.
 */
export function configureDocument({ htmlProps = {}, headChildren } = {}) {
  documentConfig = {
    htmlProps,
    headChildren,
  };
}

const Doc = NextDocument?.default ?? NextDocument;

export class Document extends Doc {
  static async getInitialProps(ctx) {
    const sheet = new ServerStyleSheet();
    const originalRenderPage = ctx.renderPage;

    try {
      ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: (App) => (props) => sheet.collectStyles(<App {...props} />),
        });

      const initialProps = await Doc.getInitialProps(ctx);
      return {
        ...initialProps,
        styles: [initialProps.styles, sheet.getStyleElement()],
      };
    } finally {
      sheet.seal();
    }
  }

  render() {
    const { htmlProps, headChildren } = documentConfig;

    return (
      <Html {...htmlProps}>
        <Head>
          {headChildren}
          <link
            rel="stylesheet"
            href="https://cdn.jsdelivr.net/npm/katex@0.16.22/dist/katex.min.css"
            integrity="sha384-5TcZemv2l/9On385z///+d7MSYlvIEw9FuZTIdZ14vJLqWphw7e7ZPuOiCHJcFCP"
            crossOrigin="anonymous"
          />
        </Head>
        <body>
          <Main />
          <Script id="theme" strategy="beforeInteractive">
            {noFlashScript}
          </Script>
          <NextScript />
        </body>
      </Html>
    );
  }
}
