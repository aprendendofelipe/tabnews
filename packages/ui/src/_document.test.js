import { render } from '@testing-library/react';
import React from 'react';

import { Document } from './_document.jsx';

describe('ui', () => {
  describe('_document', () => {
    it('renders correctly', () => {
      const { container, getByTestId } = renderOnMockRoot(<Document />);

      const bodyTag = container.querySelector('body');
      expect(bodyTag).toBeDefined();
      expect(bodyTag.innerHTML).toContain('<main');

      expect(getByTestId('Next Html')).toBeDefined();
      expect(getByTestId('Next Head')).toBeDefined();
      expect(getByTestId('Next App Content')).toBeDefined();
      expect(getByTestId('Next Scripts')).toBeDefined();

      expect(hoisted.nextScriptDefault).toHaveBeenCalledOnce();

      const themeScript = container.querySelector('script#theme');
      expect(themeScript).toBeDefined();
      expect(themeScript.innerHTML).toBe(
        `if (['auto','night'].includes(localStorage.getItem('colorMode')))
document.documentElement.setAttribute('data-no-flash', true)`,
      );

      expect(container).toMatchSnapshot();
    });

    it('should properly include styled-components styles in "initialProps"', async () => {
      const ctx = {
        defaultGetInitialProps: hoisted.getInitialProps,
        renderPage: hoisted.renderPage,
      };

      const initialProps = await Document.getInitialProps(ctx);

      expect(initialProps).toStrictEqual({
        // eslint-disable-next-line react/jsx-key
        styles: [undefined, <style>Styled Components Styles</style>],
      });

      expect(hoisted.ServerStyleSheet).toHaveBeenCalledOnce();
      expect(hoisted.getStyleElement).toHaveBeenCalledOnce();
      expect(hoisted.seal).toHaveBeenCalledOnce();

      expect(hoisted.renderPage).not.toBe(ctx.renderPage);
      expect(hoisted.renderPage).not.toHaveBeenCalledOnce();
      ctx.renderPage();
      expect(hoisted.renderPage).toHaveBeenCalledOnce();

      expect(hoisted.collectStyles).toHaveBeenCalledOnce();
    });
  });
});

const hoisted = vi.hoisted(() => {
  const nextScriptDefault = vi.fn((props) => <script {...props} />);

  const collectStyles = vi.fn();
  const getStyleElement = vi.fn(() => <style>Styled Components Styles</style>);
  const seal = vi.fn();
  const ServerStyleSheet = vi.fn(() => ({
    collectStyles,
    getStyleElement,
    seal,
  }));

  const renderPage = vi.fn((App) => collectStyles(App));
  const getInitialProps = vi.fn(() => ({}));

  return {
    nextScriptDefault,
    ServerStyleSheet,
    collectStyles,
    getStyleElement,
    seal,
    renderPage,
    getInitialProps,
  };
});

vi.mock('next/document', () => ({
  Html: vi.fn((props) => <html lang="next-i18n" data-testid="Next Html" {...props} />),
  // eslint-disable-next-line @next/next/no-head-element
  Head: vi.fn((props) => <head data-testid="Next Head" {...props} />),
  Main: vi.fn((props) => <main data-testid="Next App Content" {...props} />),
  NextScript: vi.fn((props) => <script data-testid="Next Scripts" {...props} />),
  default: class Document extends React.Component {
    static getInitialProps(ctx) {
      return ctx.defaultGetInitialProps(ctx);
    }
  },
}));

vi.mock('next/script', () => ({ default: hoisted.nextScriptDefault }));
vi.mock('styled-components', () => ({ ServerStyleSheet: hoisted.ServerStyleSheet }));

function renderOnMockRoot(ui, options) {
  const mockRoot = document.createElement('mockroot');
  const container = document.body.appendChild(mockRoot);

  return suppressDOMNestingWarnings(() =>
    render(ui, {
      container,
      ...options,
    }),
  );
}

function suppressDOMNestingWarnings(callback) {
  const originalError = console.error;
  console.error = (...args) => {
    if (args[0].includes('cannot appear as a child of') && args[1] === '<html>' && args[2] === 'mockroot') {
      return;
    }
    originalError.call(console, ...args);
  };

  try {
    return callback();
  } finally {
    console.error = originalError;
  }
}
