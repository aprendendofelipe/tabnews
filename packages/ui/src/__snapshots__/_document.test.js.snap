// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ui > _document > renders correctly 1`] = `
<mockroot>
  <html
    data-testid="Next Html"
    lang="next-i18n"
  >
    <head
      data-testid="Next Head"
    >
      <link
        crossorigin="anonymous"
        href="https://cdn.jsdelivr.net/npm/katex@0.16.22/dist/katex.min.css"
        integrity="sha384-5TcZemv2l/9On385z///+d7MSYlvIEw9FuZTIdZ14vJLqWphw7e7ZPuOiCHJcFCP"
        rel="stylesheet"
      />
    </head>
    <body>
      <main
        data-testid="Next App Content"
      />
      <script
        id="theme"
        strategy="beforeInteractive"
      >
        if (['auto','night','dark','day','light'].includes(localStorage.getItem('colorMode')))
document.documentElement.setAttribute('data-no-flash', true)
      </script>
      <script
        data-testid="Next Scripts"
      />
    </body>
  </html>
</mockroot>
`;
